### PR TITLE
Relax tag filter constraint

### PR DESCRIPTION
**Pull request by bojanserafimov**
Tags are now allowed to appear after filters, if they are on a different property of the same vertex.

Changes made:
- Removed uses of undocumented `context['tags']` and replaced them with `TagInfo` from the metadata table.
- Changed the directive processing order, so that tags at all property fields on a vertex are processed before we compile any of those property fields.

### CODE REVIEW

File: graphql_compiler/compiler/compiler_frontend.py [hunk 1]
```diff
@@ -244,6 +246,38 @@ def _process_output_source_directive(schema, current_schema_type, ast,
...
+
+    Args:
+        context: dict, various per-compilation data (e.g. declared tags, whether the current block
+                 is optional, etc.). May be mutated in-place in this function!
+        current_schema_type: GraphQLType, the schema type at the current location
+        location: Location object representing the current location in the query
+        tag_directive: GraphQL Directive that we want to process
+    """
+    if is_in_fold_scope(context):
+        raise GraphQLCompilationError(u'Tagging values within a @fold vertex field is '
```
Comment (obi1kenobi) [inline graphql_compiler/compiler/compiler_frontend.py]: Let's make sure we have a test for this. I think the module docstring's summary, if followed in the implementation, will ensure that this is handled correctly, but best to prove it with a passing test case.

File: graphql_compiler/compiler/compiler_frontend.py [hunk 2]
```diff
@@ -270,30 +304,6 @@ def _compile_property_ast(schema, current_schema_type, ast, location,
                                           .format(COUNT_META_FIELD_NAME, location))
 
     # step P-2: process property-only directives
```
Comment (obi1kenobi) [inline graphql_compiler/compiler/compiler_frontend.py]: Update this comment to the new P-2 description in the module docstring

File: graphql_compiler/compiler/filters.py [hunk 3]
```diff
@@ -113,13 +113,13 @@ def _represent_argument(directive_location, context, argument, inferred_type):
 
         return (expressions.Variable(argument, inferred_type), None)
     elif is_tagged_parameter(argument):
-        argument_context = context['tags'].get(argument_name, None)
```
Comment (obi1kenobi) [inline graphql_compiler/compiler/filters.py]: Thanks! Is anything else using `context['tags']`? If not, it's probably safe to delete.

Comment (bojanserafimov) [reply graphql_compiler/compiler/filters.py]: Nothing else. I did delete it.

File: graphql_compiler/compiler/compiler_frontend.py (hunk 1 above)
Comment (bojanserafimov) [reply graphql_compiler/compiler/compiler_frontend.py]: We do.

File: graphql_compiler/compiler/filters.py (hunk 3 above)
Comment (obi1kenobi) [reply graphql_compiler/compiler/filters.py]: Let's make sure `context['tags']` is never set up in the first place, and confirm with a grep for `context['tags']`:
https://github.com/kensho-technologies/graphql-compiler/pull/351/files#diff-fffee83da4144bc70ab3578390d57761R827

**Pull request merged**